### PR TITLE
refactor(cmake) : remove dependence of wayland when not use it

### PR DESCRIPTION
**Pull request by HR1025**

### CODE REVIEW

**Comment by HR1025** [comment]
[how]
1. move wayland related Cmake to `cmake/wayland_driver.cmake`
2. use option to choose whether build or not wayland
3. add build directory to git ignore

**Comment by HR1025** [comment]
This merge merge `CMakeLists.txt` and `wayland/CMakeLists.txt` to make them together.

Thank to @WallaceIT for previous `wayland/CMakeLists.txt`.
> Many implementations refer to it.

File: cmake/wayland_driver.cmake [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    get_filename_component(output_file_base ${protocol_xml_file} NAME_WE)
+    set(xdg_shell_file "${output_dir}/wayland-${output_file_base}-client-protocol")
+    message("-- generating ${xdg_shell_file}.h")
+    execute_process(COMMAND "${WAYLAND_SCANNER_EXECUTABLE}"  client-header  "${protocol_xml_file}" "${xdg_shell_file}.h")
+    message("-- generating ${xdg_shell_file}.c")
+    execute_process(COMMAND "${WAYLAND_SCANNER_EXECUTABLE}"  private-code "${protocol_xml_file}"  "${xdg_shell_file}.c")
+endmacro()
+
+if (NOT USE_WL_SHELL AND USE_XDG_SHELL)
```
Comment (WallaceIT) [inline cmake/wayland_driver.cmake:23]: Here you have a bug, it should be:

`if (NOT USE_WL_SHELL AND NOT USE_XDG_SHELL)`

Otherwise, enabling only XDG shell would trigger the FATAL_ERROR

File: cmake/wayland_driver.cmake [hunk 2]
```diff
@@ -0,0 +1,61 @@
+project(lv_wayland)
+
+option(USE_WL_SHELL "use wl shell (discard, no recommend)" OFF)
```
Comment (WallaceIT) [inline cmake/wayland_driver.cmake:3]: I'd rephrase the comment as:

`"use wl_shell (deprecated, not recommended)"`

File: cmake/wayland_driver.cmake [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    get_filename_component(output_file_base ${protocol_xml_file} NAME_WE)
+    set(xdg_shell_file "${output_dir}/wayland-${output_file_base}-client-protocol")
+    message("-- generating ${xdg_shell_file}.h")
+    execute_process(COMMAND "${WAYLAND_SCANNER_EXECUTABLE}"  client-header  "${protocol_xml_file}" "${xdg_shell_file}.h")
+    message("-- generating ${xdg_shell_file}.c")
+    execute_process(COMMAND "${WAYLAND_SCANNER_EXECUTABLE}"  private-code "${protocol_xml_file}"  "${xdg_shell_file}.c")
+endmacro()
+
+if (NOT USE_WL_SHELL AND USE_XDG_SHELL)
+    message(FATAL_ERROR "Please select at least one of USE_WL_SHELL and USE_XDG_SHELL (that is, can both select)")
```
Comment (WallaceIT) [inline cmake/wayland_driver.cmake:24]: I'd rephrase the comment as follows:

`"Please select at least one of USE_WL_SHELL and USE_XDG_SHELL (multiple choices allowed)"`

**Review by WallaceIT — COMMENTED** [review]
+1 for this approach. I left one bug warning and a couple of comments, but otherwise seems good to me.

Thank you!

**Comment by HR1025** [comment]
Thank you for your advice. I will revise them.

File: cmake/wayland_driver.cmake (hunk 1 above)
Comment (HR1025) [reply cmake/wayland_driver.cmake]: modified already

File: cmake/wayland_driver.cmake (hunk 2 above)
Comment (HR1025) [reply cmake/wayland_driver.cmake]: modified already

File: cmake/wayland_driver.cmake (hunk 3 above)
Comment (HR1025) [reply cmake/wayland_driver.cmake]: modified already
